For #9189 - Pocket articles support

Mugurell opened the pull request:
Since the previous version of this component was not in use anymore it was used
as a starting point for the newly needed functionality - easily offering Pocket
recommended articles to interested clients.

The new version offers a simple API for downloading, caching and refreshing in
the background the list of Pocket articles that is offered to clients.


### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

pocmo commented:
Random drive-by question. I saw you created `feature-pocket-recommendations`. There's also `service-pocket` as a generic pocket component, derived from the Amazon apps afaik. Is it possible to merge that together?

Mugurell commented:
> Random drive-by question. I saw you created `feature-pocket-recommendations`. There's also `service-pocket` as a generic pocket component, derived from the Amazon apps afaik. Is it possible to merge that together?

Not sure about the splitting and what each layer of AC should be, how should they all be stacked. I'm ok with any recommendations.
I saw `service-pocket` (based on just the naming also) as offering an abstraction over downloading (and returning) a list of items from Pocket, a general piece of functionality that could have many use cases.
Which is then used by the new `feature-pocket-recommendations` which serves a particular scope and is much more specific, imposing a certain way of getting the data from Pocket.

Having both as a single module would make things easier (for my patch) but I also see reasons for keeping these separate.
I see @mcomella  being the owner of `service-pocket`, maybe he knows more about the plans for this component, if it would be ok to be absorbed in a new one..?

mcomella commented:
(repasting something I sent to Petru b/c it's useful context here) I don’t believe any of the code is in use in `service-pocket`. They disabled the Pocket video recommendation endpoint and the other code was used for a prototype and external folks don’t have the API keys to use it. I’m fine with anything – this getting merged into your new component, your new component getting merged here, or even with this getting deprecated – we’re not actively using the code and I see no near future in which we do use it so I don’t think it’s worth effort to make sure we keep it working.

I see I’m the code owner in the service-pocket module – I don’t actually have cycles to review this. I think it’s okay for someone else to review this – I assume it’s just a normal get-and-serialize-json-data PR. The general advice I’d mention is:
- Make sure the retry code is correct, if it exists. We once had a bug where we retried many times per second, causing denial-of-service behavior and that’s really bad because we can’t guarantee users update to the latest version
- From a performance perspective, if the UI doesn’t require on-demand updates, it’s simpler and probably more effective to do this via WorkManager so that users won’t be waiting on fetches and fetches won’t slow down other requests in the app

pocmo commented:
Makes sense, thanks Mike!

I think in any case we should still end up with only one component providing Pocket functionality. If we need more than it is very likely to share code. Maybe you could turn this into the new `service-pocket` ("service" because it primarily provides access to a backend service) and we get rid of the existing code, since - as mentioned - it's not in use and unlikely to get used unfortunately.

Mugurell commented:
Yeah, thinking more about it, maybe it makes more sense for this to be just a service.
I started this new `feature-pocket..` as working similar to `feature-top-sites`.
All the business logic being in the ac, clients interested in using the "pocket feature" having to then implement the "Pocket View" interface that would ensure them adhering to a set of specific actions.
But if everything this that this component would do is just allow querying a list of pocket articles a service might make more sense.
And if in the future more functionality is to be added it can then be easily refactored in a new feature.

Mugurell commented:
Sorry this got so big and is a single commit, but should be straightforward and easy to follow.

I'd say the most interesting part is exactly the entry point to this functionality - PocketStoriesService#startPeriodicStoriesRefresh
Because transparent data refresh in the background is important and a PeriodicWorkRequest was used for this we also need to know when to stop this work.
As such what I recommended is to use PocketStoriesService#startPeriodicStoriesRefresh and PocketStoriesService#stopPeriodicStoriesRefresh as higher up in the client app as possible (preferable in the Application or the single Activity).
Open to other suggestions regarding this but testing it lightly on Fenix it seems to work ok.

jonalmeida commented on components/service/pocket/src/main/java/mozilla/components/service/pocket/api/PocketEndpoint.kt:
```diff
@@ -0,0 +1,67 @@
+         */
+        fun newInstance(
+            pocketApiKey: String,
+            client: Client = HttpURLConnectionClient()
```
Probably best to remove this default and the `lib-fetch-httpurlconnection` in the `build.gradle`.

csadilek commented on components/service/pocket/src/test/java/mozilla/components/service/pocket/api/PocketEndpointRawTest.kt:
```diff
@@ -0,0 +1,120 @@
+        defaultResponse = errorResponse
+
+        client = mock<Client>().also {
+            Mockito.`when`(it.fetch(any())).thenReturn(defaultResponse)
```
nit: let's import `Mockito.when`

Mugurell replied:
Totally!

csadilek commented on components/service/pocket/src/test/java/mozilla/components/service/pocket/api/PocketEndpointRawTest.kt:
```diff
@@ -0,0 +1,120 @@
+        assertRequestParams(client, makeRequest = {
+            endpoint.getGlobalArticlesRecommandations(TEST_ARTICLES_COUNT, TEST_ARTICLES_LOCALE)
+        }, assertParams = { request ->
+            Assert.assertEquals(expectedUrl, request.url)
```
nit: import `Assert.`

csadilek commented on components/service/pocket/src/test/java/mozilla/components/service/pocket/api/PocketEndpointTest.kt:
```diff
@@ -0,0 +1,101 @@
+
+    @Test
+    fun `WHEN getting articles recommendations and the endpoint returns null THEN a failure is returned`() {
+        Mockito.`when`(raw.getGlobalArticlesRecommandations(TEST_ARTICLES_COUNT, TEST_ARTICLES_LOCALE)).thenReturn(null)
```
nit: import `Mockito.when`. You can also use `whenever` to avoid the back ticks.

csadilek commented on components/service/pocket/src/main/java/mozilla/components/service/pocket/stories/update/RefreshPocketWorker.kt:
```diff
@@ -0,0 +1,73 @@
+            ) {
+                Result.success()
+            } else {
+                Result.failure()
```
@Amejia481 and I ran into this before and found that returning `failure` here actually prevents the periodic task from running again, which is not what you want in this case. We'd have to verify again, but then you'd want to return `retry` or `success` here instead.

Amejia481 replied:
Yeah,  you can find some docs [here](https://developer.android.com/reference/androidx/work/ListenableWorker.Result#failure()), additionally we did something for `AddonUpdaterWorker` where [we stopped a periodic work](https://github.com/mozilla-mobile/android-components/blob/a9e5357a4ea19ceb07de96b3ccf5e72a56d3b427/components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt#L591) from getting executed when add-on was not installed anymore.

Mugurell replied:
Thank you!

csadilek reviewed (changes requested):
Looks great, @Mugurell. I found a few things to fix and some nits to address below.

Two additional topics to discuss, not covered below:
- I would prefer if we didn't mix `Stories` and `Articles` as much. It was quite confusing to read, as it's all the same, but we're switching back and forth quite a bit in the API. Should we just use `Story` as a concept and skip `Article`. If not, what's the difference?
- I am not convinced we should use Room (or a sqlite database at all) to cache pocket stories. It's quite some extra complexity and we're really only querying all data at once i.e. we're not using any of the query functionality provided by the DB. It seems just storing a flat file of the last response would be much easier and efficient? Do you foresee more complicated use cases where we would need structured queries?

rocketsroger commented on docs/changelog.md:
```diff
@@ -21,6 +21,10 @@ permalink: /changelog/
 * **browser-icons**
   * Catch `IOException` that may be thrown when deleting icon caches.
 
+* **service-pocket**
```
This will have to be moved to the current A-C development version.

rocketsroger commented on components/service/pocket/src/test/java/mozilla/components/service/pocket/stories/PocketRecommendationsRepositoryTest.kt:
```diff
@@ -0,0 +1,60 @@
+            verify(dao).cleanOldAndInsertNewPocketStories(apiArticlesMappedForDb)
+        }
+    }
+}
```
Please add blank line after this.  Same as the other files.  Thanks,

rocketsroger commented on components/service/pocket/src/test/java/mozilla/components/service/pocket/helpers/Params.kt:
```diff
@@ -0,0 +1,12 @@
+internal const val TEST_ARTICLES_COUNT = 3
+internal const val TEST_ARTICLES_LOCALE = "EN-us"
+internal const val TEST_VALID_API_KEY = "55555-242424242424242424242424"
+internal val TEST_URL = "https://mozilla.org".toUri()
```
Add extra line.  Thanks

Mugurell commented on components/service/pocket/build.gradle:
```diff
@@ -22,12 +29,18 @@ android {
     implementation project(':support-ktx')
     implementation project(':support-base')
     implementation project(':concept-fetch')
+    implementation project(':lib-fetch-httpurlconnection')
```
I saw Jonathan recommending the same.
I would be curious as to why.
I wanted this component to be plug and play, as easy to use as possible and so for me it makes more sense for it to have everything it needs, have full control over it's dependencies and internal behavior while potentially allowing clients to override part of.
As a client (let's say other than Fenix) only interested in a list of Pocket articles I'd prefer to not care about how the download is made in the background and I may not even have an implementation of `concept.fetch.Client` to pass to this component.

jonalmeida replied:
To add to the above, for non-Fenix applications as well, making a conscience decision to use their own provided HTTP client will make it clear what control they have to maintain; avoids creating multiple clients unknowingly.

We provide non-GV fetch clients for non-Mozilla consumers, but _how_ that resource is instantiated and maintained _needs_ to be decided by them. This avoids creating a new client for every service/ui/feature component unintentionally as well.

Mugurell commented on components/service/pocket/README.md:
```diff
@@ -1,13 +1,17 @@
 
 ## Usage
+- Use PocketStoriesService#startPeriodicStoriesRefresh and PocketStoriesService#stopPeriodicStoriesRefresh
+as higher up in the client app as possible (preferable in the Application or the single Activity) ensuring the
```
Thank you!

Mugurell commented on components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketRecommendedArticle.kt:
```diff
@@ -0,0 +1,32 @@
+ * @param publishedTimestamp unknown: please ask for clarification if needed.
+ * @param engagement unknown: please ask for clarification if needed.
+ * @param dedupeUrl the full url to the article's page. Will contain "?utm_source=pocket-newtab".
+ * @param sortId the index of this recommendation in the list which is sorted by date added to the API results.
```
Thank you!

Mugurell commented on components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketStoriesService.kt:
```diff
@@ -0,0 +1,57 @@
+ * @param context Android Context. Prefer sending application context to limit the possibility of even small leaks.
+ * @param pocketArticlesConfig full configuration for how and what pocket articles to get.
+ */
+class PocketStoriesService(
```
Thank you, will use `Story` everywhere.

csadilek reviewed (commented):
> @csadilek Regarding serializing the stories in a File versus a db I'd prefer the db route just to get the ACID guarantees ....  synchronize read/writes in the File. We don't use any advanced queries or relations but the code is pretty clear, straight-forward and saves us from a bit of headaches.

@Mugurell Right, but we also have `AtomicFile` and ext. functions to provide that, plus we're already using that in similar instances where we fetch and cache from a server e.g. in our `AddonCollectionProvider`.

More generally, I think we should only take the overhead of a full DB if we intend to use its powers, which we may or may not want here in the future. The fact that we are currently throwing all records away and inserting/querying them all together is a smell. We're basically using the DB as a flat file here.

I will leave it up to you :)

Mugurell commented on components/service/pocket/src/main/java/mozilla/components/service/pocket/api/PocketGlobalArticleRecommendation.kt:
```diff
@@ -0,0 +1,36 @@
+// instances of this data type, will be required to implement code to migrate the persisted data. To prevent this,
+// we mark the constructors internal. If a consumer is required to persist this data, we should consider providing a
+// persistence solution.
+internal data class PocketGlobalArticleRecommendation internal constructor(
```
Maybe having just `A Pocket recommended story.` as description for the type offered to clients is enough.

Mugurell commented:
> More generally, I think we should only take the overhead of a full DB if we intend to use its powers, which we may or may not want here in the future. The fact that we are currently throwing all records away and inserting/querying them all together is a smell. We're basically using the DB as a flat file here.

Thank you!
Will look into switching to using an AtomicFile.

Mugurell commented on components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketStoriesService.kt:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * Single starting point for the "get Pocket articles" functionality.
+     *
+     * Use this at an as low as possible level in your application.
```
Guess it's a matter of perspective 😅.
Let's consider the Android Application at the top and then will use "as high as possible".

csadilek commented on components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketRecommendedArticle.kt:
```diff
@@ -0,0 +1,32 @@
+ * @param excerpt a summary of the article.
+ * @param imageSrc a url to a still image representing the article.
+ * @param publishedTimestamp unknown: please ask for clarification if needed.
+ * @param engagement unknown: please ask for clarification if needed.
```
Yeah let's remove if always empty now.

csadilek commented on components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketRecommendedArticle.kt:
```diff
@@ -0,0 +1,32 @@
+ * @param title the title of the article.
+ * @param excerpt a summary of the article.
+ * @param imageSrc a url to a still image representing the article.
+ * @param publishedTimestamp unknown: please ask for clarification if needed.
```
👍

Mugurell commented:
I've updated the PR based on your feedback, thank you all!
Did some small tests in Fenix to see that the entire flow works, all good.

Updates just pushed:
- Updated various naming issues all throughout the patch.
- Switched to using an AtomicFile to persist the stories locally, all seems to work okay and the code is simpler.
- Removed `HttpURLConnectionClient` default from code (implementation still used in tests though). Initially all configuration was injected in the Worker but there is no way to pass the Client as `extra` data to the Worker so I went with caching both the Pocket api key and the Client in `PocketStoriesUseCases`'s companion object so that when the Worker uses an usecase to download new data we'll have the passed in Client available.
- Updated / added tests accordingly.

csadilek reviewed (approved):
Thanks, @Mugurell. I found a few things to fix up before landing, but ✅ 🚢 .

Mugurell commented on components/service/pocket/src/test/java/mozilla/components/service/pocket/api/PocketJSONParserTest.kt:
```diff
@@ -0,0 +1,99 @@
+// import kotlin.reflect.KVisibility
+//
+// @RunWith(AndroidJUnit4::class)
+// class PocketJSONParserTest {
```
This was used in the initial version to convert the Pocket JSON to out Kotlin model classes.
If we're persisting the JSON this is not needed anymore but I forgot to delete the file.
Thank you for spotting it!

Mugurell commented on components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketArticlesConfig.kt:
```diff
@@ -0,0 +1,43 @@
+ *     Default value is 3.
+ * @param locale Optional - ISO-639 locale used to retrieve recommended items for a certain region.
+ */
+class PocketStoriesConfig(
```
Thank you!
Renamed it now.

Mugurell commented on components/service/pocket/build.gradle:
```diff
@@ -4,13 +4,20 @@
+
+        kapt {
+            arguments {
+                arg("room.schemaLocation", "$projectDir/schemas".toString())
```
Indeed. Thank you!

csadilek reviewed (approved):
🚢
